File: LeetCodePython/py44.py
# 通配符匹配
# ？匹配任意一个字符； * 匹配任意字符串，包含空串
# 动态规划
class Solution:
    def isMatch(self, s: str, p: str) -> bool:
        s_len = len(s)
        p_len = len(p)
        
        dp = [[False]*(p_len+1) for _ in range(s_len+1)]
        dp[0][0] = True
        
        for i in range(p_len):
            if p[i]=='*' and dp[0][i]:
                dp[0][i+1] = True
        #print(dp)
        for i in range(s_len):
            for j in range(p_len):
                first_match = p[j] in [s[i], '?']
                if p[j] == '*':
                    dp[i+1][j+1] = dp[i][j+1] or dp[i+1][j]
                else:
                    dp[i+1][j+1] = first_match and dp[i][j]
        #print(dp)
        return dp[-1][-1]

File: LeetCodeCPlusPlus/py44.cpp
// 通配符匹配
#include <string>
using namespace std;
class Solution {
public:
    bool isMatch(string s, string p) {
        int n = s.size(), m = p.size();
        bool dp[n+1][m+1];
        memset(dp, 0, sizeof(dp));  
        dp[n][m] = true;
        // 循环从n开始，为了处理串s为空时，串p能否匹配的问题
        for (int i=n;i>=0;i--)
            for (int j=m;j>=0;j--) {
                if (i==n && j==m)
                    continue;
                bool first_match = j<m && i<n &&(p[j] =='?' || p[j] == s[i] || p[j] == '*');
                if (j<m && p[j] == '*')
                    //将 * 跳过 和将字符匹配一个并且 pattern 不变两种情况; first_match做了边界处理
                    dp[i][j] = dp[i][j+1] || first_match && dp[i+1][j];
                else
                    dp[i][j] = first_match && dp[i+1][j+1];
            }
                
        return dp[0][0];
        
    }
};